esphome/components/store_yaml/decompressor.cpp

```diff
@@ -0,0 +1,125 @@
+#include "decompressor.h"
+#include "esphome/core/log.h"
+#include "esphome/core/hal.h"
+
+namespace esphome {
+namespace store_yaml {
+
+static const char *const TAG = "store_yaml";
+
+Decompressor::Decompressor(const uint8_t *ptr, size_t len) {
+  this->data_ptr_ = ptr;
+  this->data_len_ = len;
+  this->reset();
+}
+
+void Decompressor::reset() {
+  this->pos_ = 0;
+  this->size_ = 0;
+  this->buff_ = 0;
+  this->codes_.clear();
+  for (uint32_t i = 0; i < 256; i++) {
+    this->codes_.push_back(Entry{.p = 0, .c = (uint8_t) i});
+  }
+  this->code_width_ = 9;  // log2next + 1
+}
+
+bool Decompressor::is_eof() const { return this->pos_ >= this->data_len_; }
+
+uint32_t Decompressor::get_bits_(size_t bits) {
+  if (this->is_eof()) {
+    return UINT32_MAX;
+  }
+
+  while (this->size_ < bits) {
+    this->buff_ = (this->buff_ << 8) | progmem_read_byte(&this->data_ptr_[this->pos_++]);
+    this->size_ += 8;
+  }
+
+  uint32_t value = (this->buff_ >> (this->size_ - bits)) & ((1 << bits) - 1);
+  this->size_ -= bits;
+  this->buff_ &= (1 << this->size_) - 1;
+  return value;
+}
+
+const Entry *Decompressor::get_entry_(uint16_t &code) {
+  if (this->codes_.size() == ((1 << this->code_width_) - 1)) {
+    this->code_width_++;
+  }
+  code = (uint16_t) this->get_bits_(this->code_width_);
+  if (code < this->codes_.size()) {
+    return &this->codes_[code];
+  }
+  if (code != this->codes_.size()) {
+    this->pos_ = this->data_len_;  // error in input, set eof
+  }
+  return nullptr;
+}
+
+std::string Decompressor::get_string_(const Entry *entry) const {
+  std::string s(1, (char) entry->c);
+  while (entry->p != 0) {
+    if (entry->p >= this->codes_.size()) {
+      ESP_LOGE(TAG, "entry->p %d not found", entry->p);
```

Comment:
```suggestion
      ESP_LOGE(TAG, "entry->p %" PRIu16 " not found", entry->p);
```
`entry->p` is `uint16_t` type right?

Reply:
I don't really care about these, since smaller than int are promoted and passed as int.

Reply:
You're right, it will work for now. I just wanted to future-proof it. If entry->p is changed to uint32_t and the MSB gets set, it could print negative values. Similarly, on platforms where int is 2 bytes, the same issue would occur if the MSB of a uint16_t gets set. Since ESPHome is becoming more of a multi-platform project, I think it's good to account for these edge cases early.

Reply:
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/migration-guides/release-5.x/5.0/gcc.html#espressif-toolchain-changes

I dont know whats best practice here, this created unwanted headache in ESPHome as i understood it.

Reply:
This PRIu16 breaks compilation with the current Arduino framework. Will this be ever fixed?

Reply:
https://github.com/esphome/esphome/pull/8155

Does this fix it? Merge latest dev into gabest11/esphome#store_yaml

Reply:
That helped!